ssg/boolean_expression.py

```diff
@@ -68,23 +130,62 @@ def __call__(self, **kwargs):
         if self.arg:
             full_name += '[' + self.arg + ']'
         val = kwargs.get(full_name, False)
+        if self.ver_specs:
+            if type(val) is str:
+                return val in self.requirement
+            return False
         return bool(val)
 
+    def __hash__(self):
+        return hash(self.as_id())
+
+    def __eq__(self, other):
+        return hash(self) == hash(other)
+
     def __lt__(self, other):
         return self.as_id() < other.as_id()
 
     def as_id(self):
         id_str = self.name
         if self.arg:
             id_str += '_' + self.arg
+        for spec in sorted(self.ver_specs):
+            id_str += '_' + _version_specifier_to_id(spec)
         return id_str
 
     def as_dict(self):
-        res = {'id': self.as_id(), 'name': self.name, 'arg': ''}
+        res = {'id': self.as_id(), 'name': self.name, 'arg': '',
+               'ver_title': '', 'ver_cpe': '', 'ver_specs': []}
+
         if self.arg:
             res['arg'] = self.arg
+
+        if self.ver_specs:
+            res['ver_title'] = ' and '.join(['{0} {1}'.format(
+                ssg.utils.comparison_to_oval(op), ver)
+                for op, ver in self.ver_specs])
+
+            res['ver_cpe'] = ':'.join(['{0}:{1}'.format(
+                ssg.utils.escape_comparison(op), ver)
+                for op, ver in self.ver_specs])
```

Comment:
lets take `foo>1.0,<2.0`:

The `ver_specs` is a list of individual specs: `op: >, ver: 1.0, etc...`, `op: <, ver: 2.0, etc...`, whereas `ver_cpe` in that case is a composition `cpe:/a:foo:lt:1.0:gt:2.0` (yes, CPE name does not make much sense in this case but we are going to get rid of dictionaries anyway). The `ver_title` would be `greater than 1.0 and less than 2.0`.

Reply:
OK, and where are they used?

Reply:
The `ver_cpe` was designed to be used to update `name` for CPEs. I just missed that part of code because scanner does not care about uniqueness of CPE ID in dictionaries. The title was designed for things like `<criteria>` titles. I'll add it to the `package`.

Reply:
Done.